kernel/timeout.c

```diff
@@ -127,7 +127,22 @@ void z_add_timeout(struct _timeout *to, _timeout_func_t fn,
 		}
 
 		if (to == first()) {
+#if CONFIG_TIMESLICING
+			/*
+			 * This is not ideal, since it does not
+			 * account the time elapsed since the the
+			 * last announcement, and slice_ticks is based
+			 * on that. It means the that time remaining for
+			 * the next announcement can be lesser than
+			 * slice_ticks.
+			 */
+			if (_current_cpu->slice_ticks <= 0 ||
+			    next_timeout() < _current_cpu->slice_ticks) {
+				z_clock_set_timeout(next_timeout(), false);
+			}
```

Comment:
Coming back to look super-carefully...

OK, I think I understand what you're saying.  The logic here is indeed duplicating effectively identical code inside next_timeout().  But next_timeout() has a bug where it will return "slice_ticks" as its result, **forgetting** that elapsed() may not be zero due to unannounced ticks.

Are you sure you can't fix this inside next_timeout() by just subtracting ticks_elapsed from the result?  That would seem cleaner.  Absent that, you should be able to remove the older code entirely, I don't see why we need both.